### ISSUE TITLE
Provide an `exclude` option for linters in .analysis_options

### ISSUE THREAD

**Issue by georgelesica-wf** [dart-lang/linter#320]
It would be nice to be able to turn off linters, but keep the analyzer itself running on a specific directory within a project.

The immediate use case I have in mind is the `test/` directory. We try to use strong mode consistently, and would prefer to enable most of the linters. However, a lot of test code ends up failing on one linter or another. Sometimes this is due to issues that could be resolved with a few extra lines of code (we don't always cancel stream subscriptions in tests, for instance), but often there are very good reasons for the failures. For example, `expectAsync` often requires a type cast (`as`) when strong mode is on.

Would it be possible to support something like below?

``` yaml
analyzer:
  strong-mode: true
  exclude:
    - do/not/analyze/this/**

linter:
  rules:
    - so
    - many
    - linters
  exclude:
    - test/**
```

This would essentially say to allow the analyzer, but not the linters, to run on `test/`. I think that anything excluded in the `analyzer` section should also be excluded from linting, so `do/not/analyze/this/**` would be implicitly included in the second `exclude` block.

**Comment by zoechi** [comment]
I run into this requirements with code generation tools like the built_xxx (https://pub.dartlang.org/packages/built_value).
I do not have control over code generated by this package and it produces lots of linter warnings.

**Comment by alexeieleusis** [comment]
Can we please raise priority on this? Is blocking me from adding the latest rules in our CI since generated code violates rules :(

**Comment by bwilkerson** [comment]
Anything excluded from analysis will not have lints run on them.

What I'm not understanding is why you want to perform any analysis on generated code. What's the use case for doing so?

**Comment by georgelesica-wf** [comment]
@bwilkerson our use-case isn't generated code, so I can't speak to that. We are interested in preventing our tests from being linted.

**Comment by bwilkerson** [comment]
@georgelesica-wf Sorry, I didn't make it clear, but yes, I understood your use case. But the request to raise the priority is for a use case I don't understand and would like to.

**Comment by alexeieleusis** [comment]
I am running in an infrastructure shared with many teams, so I cannot change it and it analyzes generated code to verify correctness, I just talked to the team generating the code and changing the code generating tool is not a priority to them.
Since there are other needs with the same solution, I thought of chiming here as a last resort.

**Comment by zoechi** [comment]
@bwilkerson when I enable new linter rules I get tons of warnings until the used builders `built_value`, `built_redux`, `intl`, ... update their generated code.
See also dart-lang/intl#172 where the package maintainer is reluctant to update the generated code.

**Comment by zoechi** [comment]
This is becoming a real pain with built_value, intl, built_redux, ...
since the option was removed from source_gen to add common headers to generated files.
https://github.com/dart-lang/source_gen/issues/275

See also
https://github.com/google/built_value.dart/pull/523#discussion_r227485386
https://github.com/google/built_value.dart/issues/490
https://github.com/google/built_value.dart/issues/456
...

Perhaps support for
```
// ignore_for_file: all
```
would be easier to get?

Related to 

- https://github.com/dart-lang/sdk/issues/34069
- https://github.com/dart-lang/source_gen/issues/370

**Comment by pq** [comment]
> This is becoming a real pain with built_value, intl, built_redux, ...
since the option was removed from source_gen to add common headers to generated files.

It seems like  `source_gen` should really support this (for other use cases too).  Curious why it was dropped?  The issue suggests it was implemented in `PartBuilder` and `LibraryBuilder`; that doesn't do the trick?

Having said that, knowing what to ignore will continue to be a problem.

As for this idea:

    // ignore_for_file: all

one possible hitch is that `ignore`s identify analysis errors generally and presumably you only want to ignore lints?

**Comment by pq** [comment]
Chatting w/ @bwilkerson , it looks like there's a good solution on the way.  The analyzer is in the process of finalizing a new API that, among many other things, supports nested analysis options files.  Once this is final and CLI and server clients have migrated to it, we should be able to solve this problem neatly by putting an overriding analysis options file in directories whos files we don't want to lint.  Better still this new approach will allow not just exclusion but custom configuration too meaning you could enable some but not all lints, for example, for `examples` or test source.

I think this should just do it, but by all means chime in if you think this solution misses any imagined use cases.

Thanks for the continued conversation!

**Comment by pq** [comment]
Thanks all for your continued patience.  I know this feature has been long in the wings!

**Comment by alextekartik** [comment]
Will the solution proposed allow to exclude a specific rule (rather than a set of files) after including a set of rules without touching the source code or writing all the rules? Let say that I'm including pedantic but want to ignore one rule (but keep the others without having to write yet another file and likely omit some rules). Will something like this be supported?

```yaml
include: package:pedantic/analysis_options.yaml

linter:
  exclude_rules:
    - unawaited_futures
```

**Comment by pq** [comment]
@alextekartik: what you want to do is actually possible today though the API is a little odd (and possibly entirely undocumented).  As it happens the list of scalar values passed to the `rules` part of the linter configuration is actually shorthand for map entries (`- unawaited_futures` is sugar for `unawaited_futures: true`).  Taking advantage of that you could try something like:

```
include: package:pedantic/analysis_options.yaml

linter:
  rules:
    unawaited_futures: false
```

Alternatively, you can do something like this

```
analyzer:
  errors:
    unawaited_futures: ignore
```

which is awkward in its own way but if you already have an `errors` section maybe less jarring.

**Comment by alextekartik** [comment]
Thanks @pq indeed the following:

```yaml
include: package:pedantic/analysis_options.yaml

# remove some rules
linter:
  rules:
    unawaited_futures: false
```

did the trick. It definitely helps and would be worth documenting now that we have global 'pedantic rules'. If these rules evolve we can control and skip some rules when needed as an alternative/temporary solution to:

(from <https://www.dartlang.org/guides/language/analysis-options>)
> Note: When a new version of pedantic is published, code that previously passed analysis might start failing analysis. We recommend updating your code to work with the new rules. Other options are to use a specific version of pedantic or to explicitly enable individual linter rules.

**Comment by pq** [comment]
Thanks @alextekartik.

@kwalrath: maybe we could partner up on documenting this?

**Comment by DanTup** [comment]
@pq 

> I think this should just do it, but by all means chime in if you think this solution misses any imagined use cases.

I just came across this issue because I'm getting lint warnings for `json_serializable` generated files (`foo.g.dart`). My understanding of the solution proposed (nested analysis_options) is that it only covers if you want to apply rules to a whole folder, not files (eg. `*.g.dart`).

While I may be able to disable analysis entirely for those files, this means I won't see if there are errors in the generated code (which is very easy to do, since many generators - including `json_serializable` - don't check for things like name conflicts and can generate invalid code).

**Comment by pq** [comment]
@DanTup: the idea there would be to put more permissive options files in generated source dirs that would override the ones defined in the package root.  (IOW: less pedantic lints or none at all but no blanket analysis `exclude`s.)

@bwilkerson: have I got the semantics right here?

**Comment by bwilkerson** [comment]
> I just came across this issue because I'm getting lint warnings for `json_serializable` generated files (`foo.g.dart`). My understanding of the solution proposed (nested analysis_options) is that it only covers if you want to apply rules to a whole folder, not files (eg. `*.g.dart`).

Correct.

Does that imply that the generated files are being written into directories that also contain non-generated files?

> While I may be able to disable analysis entirely for those files, this means I won't see if there are errors in the generated code (which is very easy to do, since many generators - including json_serializable - don't check for things like name conflicts and can generate invalid code).

It's totally off topic, but I'm curious: How do you resolve such issues?

> The analyzer is in the process of finalizing a new API that, among many other things, supports nested analysis options files.

The phrase "in process" might be too strong a characterization. Other tasks keep getting assigned higher priority, which is why this arc of work has not yet been completed. We might need to re-think our plan to wait for that work and try to solve the problem in a more direct way.

**Comment by DanTup** [comment]
> Does that imply that the generated files are being written into directories that also contain non-generated files?

Yeah, `json_serializable` is writing `foo.g.dart` alongside my `foo.dart` (which has the annotations). I don't know if this is configurable, but it seems to be the default at least (I'm fairly sure I didn't pick it :-))

> It's totally off topic, but I'm curious: How do you resolve such issues?

Y'know, as I write this - I realise I've confused `json_serializable` with my own code-gen again :-) I have some code-gen that pulls a JSON spec from the web and generates classes, which have annotations, which `json_serializable` uses to generate the serialization code. The error was that *my* code-gen generated the `Map` class.

That said, the question is still relevant - maybe it's still possible to make `json_serializable` generated bad code with the right annotations.

For the issue above, I saw lots of errors in my Problems view for the generated file (I hadn't excluded it at the time), most along the lines of "Map doesn't have type args". I did Go-to-Def on a `Map<String, dynamic>` and saw it jump to the `Map` class that had been generated, and realised it was because of this. Then I fixed my code-gen to generate a differently named class (`TiledMap`) - though for some instances, it might've involved changing an annotation.

Although the fixes are generally in user-code, I think understanding (or even seeing them) could involve the generated files (in a perfect world, it wouldn't - and there would be some error at the source of what generates the bad code - but I'm not sure that's very feasible).

**Comment by pq** [comment]
Zeroing in on `foo.g.dart` files specifically, I wonder if we couldn't exclude those with a glob.

Does something like this even work?

```
  exclude:
    - src/**.g.dart
```

A quick look at the `package:glob` [docs](https://pub.dartlang.org/packages/glob), which we use behind the scenes, suggests it _should_?  (Sorry, haven't had a chance to verify myself!)

**Comment by DanTup** [comment]
> Does something like this even work?

I thought it would, but didn't seem to be able to, so I just ended up adding a filter to the VS Code problems view. However, the issue is that I still want to see errors in there, just not lints.

**Comment by pq** [comment]
Thanks @DanTup!  Setting aside the general issue to focus on your specifics, I'm curious about a few things.


1. For the issues with your own codegened classes, perhaps you could inject the relevant `//ignore_for_file:` headers to ignore any lints that are failing (and for whatever reason shouldn't just get fixed)?  
2. What are the lints being triggered by `json_serializable`?  Maybe they should be fixed there?

**Comment by DanTup** [comment]
1. That's a good point! I thought about that originally then got confused thinking it was json_serializable and looked for a setting to see if I could make it write it :-)
2. Right now it looks like only `prefer_expression_function_bodies`. I'm tempted to disable that one because it fires on some code I don't want to change anyway, so not a big deal. I do think some lints contradict each other though, so I'm not sure if it'd be possible to generate code that could satisfy all possibly settings of users lints.

**Comment by pq** [comment]
If you wanted to shoot for a relatively safe base set, I'd suggest the `package:pedantic` rules.

But that doesn't solve the general problem.  For that, I suggest we explore:

1. new syntax like proposed by @georgelesica-wf above that allows an `exclude` list for the linter specifically
2. fixes to our use of globs to allow for patterns like `src/**.g.dart` to work

/cc @bwilkerson

**Comment by Reprevise** [comment]
As of now, globs still don't work for things like `src/**.g.dart`?

**Comment by pq** [comment]
@Reprevise, could you try:

```yaml
  exclude:
    - 'src/**.g.dart'
```

and report back?

(Note the quotes.)

**Comment by Reprevise** [comment]
Seems to work now, I guess I was doing something wrong. Thanks, @pq.

**Comment by Reprevise** [comment]
I don't know if this is intended, but I'm in a Flutter project and it works fine at first but if I open the file, it gives me all of the warnings even though it's excluded in `analysis_options.yaml` and even after I close the file. It only goes away when I close and reopen VSCode and don't open the file.

Here's the `exclude` part:
```YAML
  exclude:
    - 'lib/**.g.dart'
    - 'lib/services/adapters/**'
```

**Comment by pq** [comment]
> I don't know if this is intended, but I'm in a Flutter project and it works fine at first but if I open the file, it gives me all of the warnings even though it's excluded in analysis_options.yaml and even after I close the file. 

Right.  I don't recall the rationale but this is a feature of the analysis server.

@bwilkerson, @scheglov: perhaps you can add some context?

**Comment by Reprevise** [comment]
I guess it makes sense but I excluded the file for a reason... is the only solution really to use the 
`// ignore for file` comment on top for all the linting errors? There should be a property in `analysis_options.yaml` to control that.

**Comment by bwilkerson** [comment]
Actually, that's a bug. The intended behavior was that diagnostics in excluded files are shown while the file is open but are removed when the file is closed. The removal of the diagnostics is, unfortunately, broken.

There has also been a lot of discussion about whether diagnostics should ever be shown for excluded files. I suspect that we should stop displaying them.

**Comment by Reprevise** [comment]
Yeah that would be great actually to just never show diagnostics for excluded files. I'm assuming there's no ETA for a fix. There's many workarounds like how VSCode has filters for the diagnostics box or using `// ignore for file` at the top but it would be nice to have that fix. Maybe that can be an option in the `analysis_options.yml` file whether or not to show diagnostics in excluded files when they're open.

**Comment by bwilkerson** [comment]
> I'm assuming there's no ETA for a fix.

No, sorry, there isn't.

> Maybe that can be an option ...

My preference would be to not have an option until we understand better what the use cases are for when it's useful to have diagnostics displayed, and then to only add an option if it really is a matter of personal preference and not something that can be accommodated automatically.

**Comment by fryette** [comment]
Is it possible disable linter analysis for few files?
I see that exclude not supported for now(VS code hint)

**Comment by pq** [comment]
@fryette, you should be able disable analysis on specified files in your options like this:

```yaml
analyzer:
  exclude:
    - lib/src/my_lib.dart
    - test/my_test.dart
```

Is that what you're after?

> I see that exclude not supported for now(VS code hint)

Where are you seeing the hint?  Could you post a screenshot?

**Comment by fryette** [comment]
@pq that was my mistake, but seems for now it impossible to disable warnings on *.g.dart files
used this code 

```
analyzer:
  exclude:
    - lib/**.g.dart
    - lib/**.chopper.dart
```

**Comment by pq** [comment]
Good deal.  Thanks for following up!

**Comment by KOGI** [comment]
I'm running into this issue now, but as far as I can tell I'm doing everything listed in this thread. Basically, I have some files generated by a 3rd-party lib ([grpc](https://pub.dev/packages/grpc)) that I would like to exclude from all linter rules. I've tried nested analysis options in the `lib/generated` dir, which seemed to be entirely ignored. I've tried adding `exclude:`  to the `linters:` section of the main analysis options file (which is invalid), and I've tried every variation I can think of for the `exclude:` option under the `analyzer:` section:

```yaml
analyzer:
  exclude:
    - lib/generated
    - lib\generated
    - 'lib/generated'
    - 'lib\generated'
    - lib\generated\
    - lib/generated/
    - lib/generated/**
    - lib/generated/**/*.dart
    - lib/generated/path/to/specific/file.dart
```
It seems that ANY mention of the `exclude:` option in this section in some manner crashes analysis, because no matter what I enter there, I suddenly get warnings/hints that are otherwise disabled across the whole project. As well as errors telling me that class names that are clearly imported (and were resolving before) are now undefined. And as soon as I remove the `exclude` section, everything starts resolving again.

Even just the following causes the flood of errors:
```yaml
analyzer:
  exclude:
    - 'asdf'
```

**Comment by pq** [comment]
Sorry for the inconvenience @KOGI.  Are you running the command line analyzer, linter binary or looking in the IDE?

**Comment by KOGI** [comment]
I'm seeing this in the IDE (IntelliJ). But I could try it on the command-line if you like?

**Comment by pq** [comment]
No need.  Just wanted to confirm that, at its heart, what you're seeing is in the analysis server.  We've got a separate issue tracking that in https://github.com/dart-lang/sdk/issues/29430. I'll give that a bump.

**Comment by KOGI** [comment]
Perfect. Thank you!

**Comment by hacker1024** [comment]
Disabling analysis disables features like auto-completion and auto-importing, which isn't pleasant for generated files, where these things are still wanted.

**Comment by pq** [comment]
@hacker1024, @georgelesica-wf, @zoechi (and others), I'd be curious if using nested options would work for your use-cases.  It hasn't been well advertised, but the analyzer does support subdirectories containing analysis options files; they essentially override options from parent directories.

For example:


```sh
my_project
  example/
    ...
  gen/
    src/
      ...
    analysis_options.yaml
  lib/
    src/
      ...
  test/
    ...
    analysis_options.yaml
  analysis_options.yaml
  pubspec.yaml
```

Here `gen` and `test` have their own options that could diverge from those defined for `my_project`.

**Comment by hacker1024** [comment]
That would be a workaround, but each rule that generated code breaks would need to manually be turned off, which is less than ideal (unless there's a file that has every rule turned off that we can include)

**Comment by pq** [comment]
Thanks for the response @hacker1024.  If you want to disable all lints via an options file, you can just leave the lint rule section out.  In fact an empty options file will do.  For example:

```
# No rules are specified to disable all linting for this directory.

# To enable specific lints, uncomment this section and add as appropriate.
# linter:
#   rules:
#     - camel_case_types
```

I'd be curious how that works for you.

**Comment by hacker1024** [comment]
That would work, thanks. There's another problem with this approach, though. Many packages (like `json_serializable` and `freezed`) generate files in the same directory as their source files. There's no way to disable linting for generated files in this case.

I'd love to be able to do something like this:

```yaml
linter:
  exclude:
    - **.g.dart
    - **.freezed.dart
```

**Comment by bwilkerson** [comment]
Actually, those patterns should work fine in the `analyzer` section's `exclude` list. Do you want them to be analyzed for everything other than lints?

**Comment by pq** [comment]
Taking @bwilkerson's suggestion to heart, would something like this work?

```
# Don't analyze generated sources. 
analyzer:
  exclude:
    - **.g.dart
    - **.freezed.dart

# And don't lint at all.

# To enable specific lints, uncomment this section and add as appropriate.
# linter:
#   rules:
#     - camel_case_types
```

**Comment by ronba** [comment]
@pq I tried using the suggestion of an empty analysis_options file - the issue that I'm seeing is that it breaks code autocompletion in vscode making generated code difficult to discover.

To reproduce the behaviour:
- Create a new flutter project: `flutter create analysis_options`
- Under the lib directory create a new `analysis` folder and under it add:
  - An empty analysis_options.yaml file.
  - A new file called `test_analysis.dart` with a simple class, for example `class TestAnalysis {}`.
- Open lib/main.dart in vscode and try using code autocompletion to find TestAnalysis (by typing TestAnalysis or using the quick fix options) - at least in my case I'm not able to find it.

I tried modifying the analysis_options file, adding an analyzer/linter - however the only way I can get autocomplete to work is if I remove the analysis_options.yaml file completely - is there a setting I'm missing?

I tested this on the stable/beta branches and also when toggling "dart.previewLSP" off/on.

**Comment by bwilkerson** [comment]
@pq is on vacation this week so he probably won't see this information until next week.

The issue you're seeing with code completion sounds like a bug to me.

@scheglov My best guess at this point is that the available declarations support isn't taking nested contexts into account. We might need to differentiate between a context created for a `package_config.json` file and one created for an `analysis_options.yaml` file in order to fix this. Either that or we decide that we're not going to support the use of analysis options files to override options in subdirectories and treat every options file as if it declared the root of a separate package (which I think would be suboptimal given that we already have a different way to declare the root of a package).

**Comment by DanTup** [comment]
I think that might be the same issue described here:

https://github.com/Dart-Code/Dart-Code/issues/2470#issuecomment-630948268

Although the related code has moved since and is now here:

https://github.com/dart-lang/sdk/blob/62d0d5ff4d1a26010382d790c49bbf31a8b6be94/pkg/analyzer/lib/src/lint/analysis.dart#L273-L279

**Comment by ronba** [comment]
Thanks @DanTup that looks like the issue I'm experiencing.
For what it's worth - the analysis_options.yaml solution still works on generated code that uses the `part` option without breaking autocomplete.

In my case I have a builder that generates dart code from proto files using [protoc_plugin](https://pub.dev/packages/protoc_plugin). 
Adding the empty analysis_options.yaml or excluding the files from the analyzer breaks discoverability of the generated code.

The two solutions I came up with to eliminate lint issues and maintain autocomplete are below. 
Both work but feel very hacky and not very friendly for developers:
1) In the builder manually edit the generated code to include additional lint ignores - that's what I'm doing now.
2) Barrel export:
    - Add a builder to generate a barrel file (for example `export.dart`) that exports all of generated dart files.
    - In the project root folder exclude `- "**/*.pb*.dart"` from the analyzer.

It seems many Flutter projects rely on code generation provided through third party packages.
It also seems likely that the code these packages generate would not always meet the same linting requirements as the importing project. Requesting the package maintainer to add additional lint ignores might work in some cases but not in others.

At least in my case, either the exclude option for linters as originally requested or fixing the issue @DanTup mentioned would address it.

**Comment by ronba** [comment]
Quick correction:
> the analysis_options.yaml solution still works on generated code that uses the part option without breaking autocomplete.

Apologies - that was a false statement. That's the issue @DanTup mentioned above.

**Comment by pq** [comment]
Check out https://github.com/dart-lang/sdk/issues/46957 which proposes a means for lints to be disabled per-file with a leading comment like:

```dart
// ignore_for_file: severity:lints

...
```

This *could* be used to disable lints in tests but is especially targeted to code generators.

Feedback welcome!

**Comment by hacker1024** [comment]
dart-lang/sdk#46957 is included in Dart 2.15. This suits my needs, although I like the idea of including this functionality in `anaylsis_options.yaml` as well.

**Comment by TheManuz** [comment]
I have still an edge case with generated code (the code generated by **flutter_intl**, specifically).

I want to exclude it from **linter**, but not from **analyzer** (because excluding it from analyzer prevent the auto-import quick action in VsCode).

I cannot edit the file, since it's generated.

This prevents me from adding custom lint rules, because the generated files doesn't exclude them.

**Comment by ronba** [comment]
@TheManuz I ran into a similar issue with a few different packages, the workaround I came up with is to:

1) Exclude the file from the analyzer
2) Create a second file in the same directory 
3) Export the generated file from the file in (2)

For example, if `router.gr.dart` had linting errors, I'd exclude it from the analyzer, create a router.dart file and add `export 'router.gr.dart';`

This avoids all lint errors and you still get autocomplete.

I realize it doesn't really scale, but if you just have a few files where you need to do this - it works.

Hopefully this workaround is the intended behaviour and won't get patched away anytime soon :)

**Comment by timshadel** [comment]
Since I came to this trying to turn off lints for generated files, I built upon @hacker1024's [comment](https://github.com/dart-lang/linter/issues/320#issuecomment-992069193) using the [`source_gen` documentation](https://github.com/dart-lang/source_gen/tree/master/source_gen#configuring-combining_builder-ignore_for_file) to arrive at this.

**build.yaml**
```yaml
targets:
  $default:
    builders:
      source_gen:combining_builder:
        options:
          ignore_for_file:
            - "type=lint"
```

Still not as clean as package-wide list of ignores for the linter, but it works for anything built on `source_gen` (like JsonSerializable, in my case).
